nexus/src/db/subnet_allocation.rs

```diff
@@ -490,142 +305,585 @@ impl QueryFragment<Pg> for FilterConflictingVpcSubnetRangesQueryValues {
     }
 }
 
+/// Errors related to inserting or attaching a NetworkInterface
+#[derive(Debug)]
+pub enum NetworkInterfaceError {
+    /// The instance specified at create or attach is associated with a
+    /// different VPC
+    InstanceSpansMultipleVpcs(Uuid),
+    /// There are no available IP addresses in the requested subnet
+    NoAvailableIpAddresses,
+    /// An explicitly-requested IP address is already in use
+    IpAddressNotAvailable(std::net::IpAddr),
+    /// Any other error
+    External(external::Error),
+}
+
+impl NetworkInterfaceError {
+    /// Construct a `NetworkInterfaceError` from a database error
+    ///
+    /// This catches the various errors that the `InsertNetworkInterfaceQuery`
+    /// can generate, especially the intentional errors that indicate either IP
+    /// address exhaustion or an attempt to attach an interface to an instance
+    /// that is already associated with another VPC.
+    pub fn from_pool(
+        e: async_bb8_diesel::PoolError,
+        interface: &IncompleteNetworkInterface,
+    ) -> Self {
+        use crate::db::error;
+        use async_bb8_diesel::ConnectionError;
+        use async_bb8_diesel::PoolError;
+        use diesel::result::DatabaseErrorKind;
+        use diesel::result::Error;
+        const MULTIPLE_VPC_ERROR_MESSAGE: &str =
+            r#"could not parse "" as type uuid: uuid: incorrect UUID length: "#;
+        const IP_EXHAUSTION_ERROR_MESSAGE: &str =
+            r#"null value in column "ip" violates not-null constraint"#;
+        const IP_NOT_AVAILABLE_CONSTRAINT: &str =
+            "network_interface_subnet_id_ip_key";
+        const NAME_CONFLICT_CONSTRAINT: &str =
+            "network_interface_instance_id_name_key";
+        match e {
```

Comment:
I wonder if any of this belongs in nexus/src/db/error.rs, and whether you can reuse any of the functions there.  Mostly there's just some details about the structure of these errors that seems tricky and it'd be nice to not duplicate it.  But I don't immediately see how to do that without refactoring those functions further.

Reply:
Yeah, I'm happy to move this around.

I am also a bit leery of this, to be honest. These error messages are extremely coupled to the database, and it's very easy to get them wrong or forget to update them as we move indexes around. I've already done that once in the process of this PR: I moved a unique index on `(vpc_id, name)` to one on `(instance_id, name)`, and forgot to update the constraint name. I really wish there were a better way to do this, but maybe it just comes with the territory of trying to catch lots of different errors from the database in one query.

I think one way to refactor usefully might be to check that the database failed a request because it violated a given constraint. For example, we could define an enum with the error variants we want to catch, along with the name of the constraint (or other error condition) that should be derived from. Then we can have a `from_pool` function which is basically this match expression, just looking for those specific strings and translating to the enum variant.

I have a bug-fix PR lined up where I do this whole song-and-dance again. I'll look at it through this lens, to see if a trait or other shared code would work.

Reply:
I'm going to hold off moving this around for this PR, but I'll prototype how to make this easier in the bug-fix PR I'm also working on.

Reply:
I've experimented with this for a bit. It can be made to work, but I'm really not sure it's worth it. Specifically, I don't think it reduces the complexity -- the solution I have in mind is definitely more opaque to me than the original. It does allow us to package the `diesel::result::DatabaseErrorKind`, string we want to check for, and the "handler" for building our error type from the database error and the candidate record. But it's pretty complicated. I'm happy to show you offline, or post it here, whatever works. I'll mull it over some more tomorrow, and maybe add it to `omicron_nexus::db::error` if it seems useful.